### PR TITLE
rgw/sfs: use sqlite for metadata, filesystem for data

### PR DESCRIPTION
**Pull request by jecluis**
Fixes: https://github.com/aquarist-labs/s3gw-core/issues/90

Signed-off-by: Joao Eduardo Luis \<joao@suse.com>

### CODE REVIEW

**Review by 0xavi0 — CHANGES_REQUESTED** [review]
I think the only important I thing I've found so far is a missing lock to the shared mutex in one of the versioned objects call.
Looks good to me, the rest of my comments are minor things.

Smoke tests pass.
Benchmark tests pass.

**Comment by 0xavi0** [comment]
Oh, I forgot to mention... Maybe squashing commits into 1 single would be also good (also minor stuff)

File: src/rgw/store/sfs/types.cc [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    ref->meta.etag = obj.etag;
+    ref->meta.mtime = obj.mtime;
+    ref->meta.set_mtime = obj.set_mtime;
+    ref->meta.delete_at = obj.delete_at;
+    ref->meta.attrs = obj.attrs;
+    objects[obj.name] = ref;
+  }
+}
+
+} // ns rgw::sal::sfs
```
Comment (0xavi0) [inline src/rgw/store/sfs/types.cc:70]: no newline here...

File: src/rgw/store/sfs/types.h [hunk 2]
```diff
@@ -162,10 +165,18 @@ class Bucket {
...
 
+using MetaBucketsRef = std::shared_ptr<sqlite::SQLiteBuckets>;
+
+static inline MetaBucketsRef get_meta_buckets(sqlite::DBConnRef conn) {
+  return std::make_shared<sqlite::SQLiteBuckets>(conn);
+}
+
 }  // ns rgw::sal::sfs
 
 #endif // RGW_STORE_SFS_TYPES_H
```
Comment (0xavi0) [inline src/rgw/store/sfs/types.h:182]: no new line

Also... the ObjectState enum could be placed in this file and remove the extra file just for it... 
I think it makes sense, as this file would have all the types used by `sfs`

**Review by 0xavi0 — COMMENTED** [review]
Sorry, pressed the wrong button and all my comments didn't get published

File: src/rgw/store/sfs/sqlite/dbconn.h [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+          sqlite_orm::make_column("object_state", &DBVersionedObject::object_state),
+          sqlite_orm::foreign_key(&DBVersionedObject::object_id).references(&DBObject::object_id))
+  );  
+}
+
+using Storage = decltype(_make_storage(""));
+
+class DBConn {
+
+  ceph::mutex lock = ceph::make_mutex("conn_lock");
```
[the remark this answers is not in the document]
Comment (jecluis) [reply src/rgw/store/sfs/sqlite/dbconn.h]: removed

File: src/rgw/store/sfs/types.h (hunk 2 above)
Comment (jecluis) [reply src/rgw/store/sfs/types.h]: While I agree with your point, I think this is something we can handle when we work on versioning.

File: src/rgw/store/sfs/sqlite/sqlite_versioned_objects.cc [hunk 4]
```diff
@@ -32,29 +32,33 @@ std::optional<DBOPVersionedObjectInfo> SQLiteVersionedObjects::get_versioned_obj
...
-  auto storage = get_storage();
+  std::unique_lock l(conn->rwlock);
+  auto storage = conn->get_storage();
   auto uuid = object_id.to_string();
   return storage.select(&DBVersionedObject::id, where(c(&DBVersionedObject::object_id) = uuid));
 }
 
 std::optional<DBOPVersionedObjectInfo> SQLiteVersionedObjects::get_last_versioned_object(const uuid_d & object_id) const {
-  auto storage = get_storage();
+  auto storage = conn->get_storage();
```
[the remark this answers is not in the document]
Comment (jecluis) [reply src/rgw/store/sfs/sqlite/sqlite_versioned_objects.cc:65]: should actually be an `std::shared_lock`, thanks for pointing it out :)

**Comment by jecluis** [comment]
> Maybe squashing commits into 1 single would be also good (also minor stuff)

I don't agree with that. Most of these are logical changes. Squashing the commits would produce one ginormous commit, and we have plenty of those already.

**Review by 0xavi0 — APPROVED** [review]
LGTM

**Pull request merged**